Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -203,146 +203,308 @@ function xASL_wrp_ResampleASL(x)
-			if mod(size(ASL_im,4), blockSize)
-				% Number of total volumes cannot be divided by blockSize
-				error(['Total number of volumes ' xASL_num2str(size(ASL_im,4)) ' cannot be composed of blocks of size ' xASL_num2str(blockSize)]);
-			else
-				PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), blockSize); % preallocate PWI
-				for iBlock = 1:blockSize
-					PWI(:,:,:,iBlock) = xASL_stat_MeanNan(ASL_im(:,:,:,iBlock:blockSize:end), 4); % Averaged PWI4D across repetitions
-				end
-			end
-        else % if Hadamard Block size == unique PLDs * number of multi-TEs -> a single repetitions -> we don't want to average
-            PWI = ASL_im;
+        PWI4D = xASL_quant_HadamardDecoding(PathASL4D{iSpace}, x.Q);
+		nVolumes = size(PWI4D,4);
+
+        %% 2) Create PWI3D
+		% Calculate Hadamard block size (number of unique PLDs.*labeling durations.* echo times) = number of volumes per repetition
+        
+
+        %% 1) EchoTime vectors (identical to below)
+        % First we deal with a too long vector (e.g., cutting off control volumes for Hadamard)
+        nTE = length(x.EchoTime);
+        if nTE>nVolumes
+            EchoTime_PWI4D = x.EchoTime(1:nVolumes); % can we do this based on the ASL4Dcontext.tsv ?
+        else
+            EchoTime_PWI4D = x.EchoTime;
+        end
+        nTE = length(EchoTime_PWI4D);
+
+        % Then we deal with too short vectors (e.g., repetitions in the case of single-PLD)
+        factorTE = nVolumes/nTE;
+        EchoTime_PWI4D = repmat(EchoTime_PWI4D(:), [factorTE 1]);
+        fprintf('%s\n', ['EchoTime vector: ' xASL_num2str(EchoTime_PWI4D)]);
+        uniqueTE = uniquetol(EchoTime_PWI4D, 0.001); % this needs to be moved to the EchoTime vector 
+        % instead of nUniqueTE which we don't use anymore
+
+
+        % % nUniqueTE = length(uniqueTE); % Obtain the number of echo times
+        % % % We do this here now for each sequence, but we could also do this
+        % % % at the start of the ASL module
+        % % 
+        % % iUniqueTE = 1:nTEs;
+        % % iUniqueTE = repmat(iUniqueTE, [1 nVolumes/nTEs]);
+
+
+        
+        % % % %% 2) PLD vectors
+        % % % unique_InitialPLD = unique(x.Q.Initial_PLD);
+        % % % 
+        % % % % with time encoded, we always skip the latest PLD
+        % % % % because that is a dummy PLD (it is in reality the control
+        % % % % scan)
+        % % % unique_InitialPLD = unique_InitialPLD(1:end-1);
+        % % % nUnique_InitialPLD = length(unique_InitialPLD);
+        % % % 
+        % % % for iPLD=1:nUnique_InitialPLD
+        % % %     startIndex = (iPLD-1).*nTEs+1;
+        % % %     endIndex = iPLD.*nTEs;
+        % % %     iUnique_InitialPLD(startIndex:endIndex) = iPLD;
+        % % % end
+        % % % iUnique_InitialPLD = repmat(iUnique_InitialPLD, [1 nVolumes/length(iUnique_InitialPLD)]);
+
+
+        %% 2) PLD vectors (identical to below)

```

Comment:
Here, it goes wrong for Hadamard, since the PLD order is not correct.
If I take in the HAD4 example the full 64 volumes and remove the end to get to 48 volumes (as you stated I should do), I get this order of PLDs:

3600,3600,3600,3600,3600,3600,3600,3600,600,600,600,600,600,600,600,600,1600,1600,1600,1600,1600,1600,1600,1600,2600,2600,2600,2600,2600,2600,2600,2600,3600,3600,3600,3600,3600,3600,3600,3600,600,600,600,600,600,600,600,600

So is this perhaps because x.Q.InitialPLD is still in the pre-decoding order?
It would be great to ensure that `xASL_quant_HadamardDecoding` also adds a decoding index vector to translate the old vectors to the post-decoding vectors.
Alternatively, you could decode the pre-existing EchoTime, PLD, LabDur vectors inside this function as well.

Reply:
Is this still relevant? You can probably resolve that.